### PR TITLE
DAOSGCP-83 and DAOSGCP-84 Automatic format storage with pool and container creation that supports ACLs

### PR DESCRIPTION
**Pull request by lsitkiew**

### CODE REVIEW

File: terraform/examples/daos_cluster/terraform.tfvars.perf.example [hunk 1]
```diff
@@ -26,11 +26,19 @@ server_daos_scm_size       = 45
...
+#   {
+#     label = "test_pool"
+#     size = "1TB"
+#     tier_ratio = 3
+#     properties = "reclaim:disabled"
+#     containers = [
+#       {
+#         label = "test_cont",
+#         type = "POSIX",
+#         properties = "rf:0"
```
Comment (markaolson) [inline terraform/examples/daos_cluster/terraform.tfvars.perf.example:38]: Right here we need to be able to support an array of user attributes on containers.  At a minimum we need a user attribute named "automount". 

```
user_attributes = [
  automount = true,
  foo=bar
]
```

Comment (cboneti) [inline terraform/examples/daos_cluster/terraform.tfvars.perf.example:38]: `properties = "rf:0"`  Would it be possible to clarify why this is needed and being provided in all examples? what does it mean? Should people always provide this? or something else instead?

File: terraform/modules/daos_server/main.tf [hunk 2]
```diff
@@ -60,17 +60,12 @@ locals {
...
     "${path.module}/templates/daos_startup_script.tftpl",
     {
+      servers           = local.servers
       first_server      = local.first_server
       daos_ca_secret_id = local.daos_ca_secret_id
       allow_insecure    = local.allow_insecure
-    }
-  )
-
-  configure_daos_content = templatefile(
```
Comment (cboneti) [inline terraform/modules/daos_server/main.tf:69]: I believe you still need a configure daos script. How are you going to start the daos daemon on clients?

**Review by cboneti — COMMENTED** [review]
I believe you still need to provide a means by which clients who do not have a daos image can mount daos. 

In a classic NFS world. You would need two outputs: install-nfs (installing nfs-common drivers) and (mount). 
In DAOS, you would need to:
1. install DAOS rpms
2. copy all the yamls
3. launch daos services

Then later, mount daos using dfuse.

File: terraform/examples/daos_cluster/terraform.tfvars.perf.example (hunk 1 above)
Comment (Kmannth) [reply terraform/examples/daos_cluster/terraform.tfvars.perf.example:38]: rf:0 is the default replication factory.   R = Replication :  F = Factory :  0 = No replica. 

 @lsitkiew   I would rather not mix Pool type changes into this PR to move these conversations to GCPDASOS-83 work. 
Also rf:0 should be set as a default value and only users what want something else need to be exposed to it.

File: terraform/modules/daos_server/main.tf (hunk 2 above)
Comment (Kmannth) [reply terraform/modules/daos_server/main.tf:69]: @lsitkiew please put back the basic configure daos script.  It is used to start Clients in some configs.

File: terraform/examples/daos_cluster/variables.tf [hunk 3]
```diff
@@ -156,9 +156,15 @@ variable "server_pools" {
...
-    containers = list(string)
+    label = string
+    size  = string
+    tier_ratio = string
+    properties = string
+    containers = list(object({
+      label = string
+      type = string
+      properties = string
+      }))
```
Comment (Kmannth) [inline terraform/examples/daos_cluster/variables.tf]: @lsitkiew  Please leave these types of changes to the GCPDAOS83  ticket.

File: terraform/examples/io500/config/config.sh [hunk 4]
```diff
@@ -30,13 +30,28 @@ ID=""
...
+    properties = "reclaim:disabled"
+    containers = [
+      {
+        label = "io500_cont",
+        type = "POSIX",
+        properties = "rf:0"
+      }
+    ]
+  }
+]'
```
Comment (Kmannth) [inline terraform/examples/io500/config/config.sh:47]: same leave for 83.

File: terraform/examples/io500/config/config.sh [hunk 5]
```diff
@@ -30,13 +30,28 @@ ID=""
...
 
 # Server(s)
 DAOS_SERVER_INSTANCE_COUNT="1"
-DAOS_SERVER_MACHINE_TYPE=n2-highmem-32 # n2-custom-20-131072 n2-custom-40-262144 n2-highmem-32 n2-standard-2
+DAOS_SERVER_MACHINE_TYPE=n2-standard-16 # n2-custom-20-131072 n2-custom-40-262144 n2-highmem-32 n2-standard-2
```
Comment (Kmannth) [inline terraform/examples/io500/config/config.sh:51]: Please put io500 config changes into a separate ticket.

File: terraform/examples/io500/config/config.sh [hunk 6]
```diff
@@ -87,6 +102,7 @@ export TF_VAR_server_machine_type="${DAOS_SERVER_MACHINE_TYPE}"
 export TF_VAR_server_os_project="${TF_VAR_project_id}"
 export TF_VAR_server_os_family="daos-server-io500-centos-7"
 export TF_VAR_server_gvnic="${DAOS_SERVER_GVNIC}"
+export TF_VAR_server_pools="${DAOS_POOLS}"
```
Comment (Kmannth) [inline terraform/examples/io500/config/config.sh:105]: This parts seems ok for 84.

File: terraform/modules/daos_server/templates/daos_startup_script.tftpl [hunk 7]
```diff
@@ -51,8 +51,8 @@ $${DAOS_DIR}/cert_gen/sm_get_ca.sh "${daos_ca_secret_id}" "server"
 systemctl enable daos_server
 systemctl start daos_server
 
-#give time for dmg commands to work 
-sleep 5
+systemctl enable daos_agent
```
Comment (Kmannth) [inline terraform/modules/daos_server/templates/daos_startup_script.tftpl:54]: @lsitkiew  There can be a race here.   Please add the sleep back in so DMG always works.

File: terraform/modules/daos_server/templates/daos_startup_script.tftpl [hunk 8]
```diff
@@ -65,40 +65,21 @@ if [[ "$${HOSTNAME}" == "$${FIRST_DAOS_SERVER_HOSTNAME}" ]]; then
...
-  do
-    if [[ $(dmg system query -v | grep Joined | wc -l) -eq ${instances} ]]; then
-      printf "\n%s\n" "DAOS storage format finished"
-      dmg system query -v
-      break
-    fi
-    printf "%s" "."
-    sleep 5
-  done
-
```
Comment (Kmannth) [inline terraform/modules/daos_server/templates/daos_startup_script.tftpl]: Please add the wait for format logic back in.

File: terraform/modules/daos_server/variables.tf [hunk 9]
```diff
@@ -146,9 +146,15 @@ variable "pools" {
...
-    containers = list(string)
+    label = string
+    size  = string
+    tier_ratio = string
+    properties = string
+    containers = list(object({
+      label = string
+      type = string
+      properties = string
+      }))
```
Comment (Kmannth) [inline terraform/modules/daos_server/variables.tf]: Same please put 83 type changes into the 83 workstream.

**Review by Kmannth — CHANGES_REQUESTED** [review]
@lsitkiew  There are a few extra changes in this ticket for 84. 

1. Please put changes to pool types into an 83 workstreams.   83 will being in a full complete data type with all the things.   Better to get this landed and build 83 work on top. 
2. I am not sure about the move to put the daso_agent starting in the client startup script.  I think there is logic today that calls as is today in some use cases.    Can we revert this change for now?   or ???? 
3. Io500 autopool seems like a good idea.  Can we wait for 83 to fully roll out all the syntax default pools need to go in the examples and io500 and keep this pull request just to the functionality?  There are default policy to discuss. 
4. Please put io500 server instances changes in their own ticket so they can be reviewed in context.   

This patch looks close thanks for the quick rebase.

File: terraform/examples/daos_cluster/terraform.tfvars.perf.example (hunk 1 above)
Comment (lsitkiew) [reply terraform/examples/daos_cluster/terraform.tfvars.perf.example:38]: Ok, I will move this changes to other PR.

File: terraform/examples/daos_cluster/variables.tf (hunk 3 above)
Comment (lsitkiew) [reply terraform/examples/daos_cluster/variables.tf]: Ok, I will move this changes to other PR.

File: terraform/examples/io500/config/config.sh (hunk 4 above)
Comment (lsitkiew) [reply terraform/examples/io500/config/config.sh:47]: Ok, I will move this changes to other PR.

File: terraform/examples/io500/config/config.sh (hunk 5 above)
Comment (lsitkiew) [reply terraform/examples/io500/config/config.sh:51]: Ok, I will move this changes to other PR.

File: terraform/modules/daos_server/main.tf (hunk 2 above)
Comment (lsitkiew) [reply terraform/modules/daos_server/main.tf:69]: I analyzed the scripts and this functionality was moved to startup script on client side: https://github.com/daos-stack/google-cloud-daos/blob/develop/terraform/modules/daos_client/templates/daos_startup_script.tftpl#L29-L30
Besides this, that file was not used by Terraform in any of the module, so that why I decided to deleted it. 
I didn't have any issue with not started DAOS agent.

File: terraform/modules/daos_server/templates/daos_startup_script.tftpl (hunk 7 above)
Comment (lsitkiew) [reply terraform/modules/daos_server/templates/daos_startup_script.tftpl:54]: I didn't get any problem here. Everything is working correctly without that sleep phase for me. I will check this.

File: terraform/modules/daos_server/templates/daos_startup_script.tftpl (hunk 8 above)
Comment (lsitkiew) [reply terraform/modules/daos_server/templates/daos_startup_script.tftpl]: I noticed that in current DAOS 2.0.2 release, command dmg system query -v is waiting for the servers to be in "Joined" state. So there is no need to add special wait functionality like this. @johannlombardi could you confirm this?

File: terraform/modules/daos_server/variables.tf (hunk 9 above)
Comment (lsitkiew) [reply terraform/modules/daos_server/variables.tf]: Ok, I will move this changes to other PR.

File: terraform/modules/daos_server/templates/daos_startup_script.tftpl (hunk 8 above)
Comment (Kmannth) [reply terraform/modules/daos_server/templates/daos_startup_script.tftpl]: @lsitkiew  per the chat we are ok to remove it.

**Comment by lsitkiew** [comment]
We decided to combine changes from DAOSGCP-83 and DAOSGCP-84 into this single PR.

File: terraform/modules/daos_server/variables.tf [hunk 10]
```diff
@@ -142,13 +142,28 @@ variable "preemptible" {
   type        = string
 }
 
+variable "pool_cont_create" {
```
Comment (cboneti) [inline terraform/modules/daos_server/variables.tf:145]: This seems like an orthogonal knob with server_pools, right?
If server_pools = [] then this means false, if server_pools is populated, then this should be automatically true, or am I missing something? Why do you need this variable?

File: terraform/examples/daos_cluster/terraform.tfvars.perf.example [hunk 11]
```diff
@@ -26,11 +26,34 @@ server_daos_scm_size       = 45
...
+#   {
+#     name = "test_pool",
+#     size = "1TB",
+#     tier_ratio = 3
+#     acls = [
+#       "A::OWNER@:rwdtTaAo",
+#       "A:G:GROUP@:rwtT"
+#     ],
+#     properties = [
+#       "reclaim:disabled"
```
Comment (Kmannth) [inline terraform/examples/daos_cluster/terraform.tfvars.perf.example:38]: No.  Do not set any non dev benchmark pool to reclaim:disabled

File: terraform/examples/daos_cluster/terraform.tfvars.tco.example [hunk 12]
```diff
@@ -26,11 +26,34 @@ server_number_of_instances = 4
...
+#   {
+#     name = "test_pool",
+#     size = "1TB",
+#     tier_ratio = 3
+#     acls = [
+#       "A::OWNER@:rwdtTaAo",
+#       "A:G:GROUP@:rwtT"
+#     ],
+#     properties = [
+#       "reclaim:disabled"
```
Comment (Kmannth) [inline terraform/examples/daos_cluster/terraform.tfvars.tco.example:38]: same

File: terraform/examples/daos_cluster/terraform.tfvars.perf.example [hunk 13]
```diff
@@ -26,11 +26,34 @@ server_daos_scm_size       = 45
...
-#    pool_size = "1TB"
-#    containers = [ "test_container1", "test_container2" ]
-#  }
+#   {
+#     name = "test_pool",
+#     size = "1TB",
+#     tier_ratio = 3
+#     acls = [
+#       "A::OWNER@:rwdtTaAo",
+#       "A:G:GROUP@:rwtT"
```
Comment (Kmannth) [inline terraform/examples/daos_cluster/terraform.tfvars.perf.example:35]: It would be better to show a actual user name in this place.  It is not clear to me where to add my user name.

File: terraform/examples/daos_cluster/variables.tf [hunk 14]
```diff
@@ -175,11 +175,19 @@ variable "server_pools" {
...
+    size       = string
+    tier_ratio = number
+    acls       = list(string)
+    properties = list(string)
+    containers = list(object({
+      name            = string
+      type            = string
+      acls            = list(string)
+      properties      = list(string)
+      user_attributes = list(string)
```
Comment (Kmannth) [inline terraform/examples/daos_cluster/variables.tf:188]: These value should all get io500 posix class defaults.

File: terraform/examples/only_daos_server/variables.tf [hunk 15]
```diff
@@ -155,11 +155,19 @@ variable "server_pools" {
...
+    acls       = list(string)
+    properties = list(string)
+    containers = list(object({
+      name            = string
+      type            = string
+      acls            = list(string)
+      properties      = list(string)
+      user_attributes = list(string)
+    }))
+  }))
```
Comment (Kmannth) [inline terraform/examples/only_daos_server/variables.tf:170]: https://github.com/daos-stack/google-cloud-daos/pull/40/files#diff-636c2c61569c377aa9d379a373d2cd9ae6b72dc0ef8d9c96a983f04c38d2142c  lets just have 1 "Type" delectation for Pools having 2 places of the same thing will be confusing in a very sort time.

File: terraform/modules/daos_server/templates/startup_script.tftpl [hunk 16]
```diff
@@ -38,4 +38,6 @@ systemctl start daos_server
 systemctl enable daos_agent
```
Comment (Kmannth) [inline terraform/modules/daos_server/templates/startup_script.tftpl:38]: Fine I will retest.  My testing need a few seconds of sleep on a single server for DMG commands to not error out.

File: terraform/modules/daos_server/variables.tf [hunk 17]
```diff
@@ -170,9 +170,17 @@ variable "pools" {
...
+    acls       = list(string)
+    properties = list(string)
+    containers = list(object({
+      name            = string
+      type            = string
+      acls            = list(string)
+      properties      = list(string)
+      user_attributes = list(string)
+    }))
+  }))
```
Comment (Kmannth) [inline terraform/modules/daos_server/variables.tf:185]: This is a 3rd spot for this list.

File: terraform/modules/daos_server/templates/pool_cont_create.tftpl [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+  do
+    sleep 5
+  done
+  echo "All DAOS Servers started"
+
+  echo "Formatting storage on servers: ${servers}"
+  dmg storage format
+  dmg system query -v
+  echo "Done formating DAOS server"
+
```
Comment (Kmannth) [inline terraform/modules/daos_server/templates/pool_cont_create.tftpl:18]: Only the below if for creating pools and containers. The above sections is for "formatting" the servers.   Likely the above should go back into the server startup section.

**Review by Kmannth — CHANGES_REQUESTED** [review]
Are 2  major issue: 
1. Showing  "reclaim:disabled" to any user is a no-go for me. 
  This is a DEV only feature of DAOS. 
2.  There new pool type is in 3? spots.   Lets find a way to have a single pool type. 
3. There is a new script called pool_cont_create  but it is also "Formats" daos. 

Minor issue: 
After reading the patch I don't know how to use the ACL feature.   I think I add my user name  in some spots but I will have to read the DAOS manual. 

There are also sever another things outside of 84/83 in this patch but lets just leave them here for now...

File: terraform/examples/daos_cluster/terraform.tfvars.perf.example (hunk 13 above)
Comment (Kmannth) [reply terraform/examples/daos_cluster/terraform.tfvars.perf.example:35]: Maybe a comment with example would be better?

**Review by Kmannth — DISMISSED** [review]

File: terraform/modules/daos_server/templates/pool_cont_create.tftpl (hunk 18 above)
Comment (Kmannth) [reply terraform/modules/daos_server/templates/pool_cont_create.tftpl]: I have a patch in progress to put this into a daos_server_format.

File: terraform/modules/daos_server/templates/startup_script.tftpl (hunk 16 above)
Comment (Kmannth) [reply terraform/modules/daos_server/templates/startup_script.tftpl:38]: Please put the 5 seconds sleep back.  dmg cannot talk for a few seconds and the logic of startup gets messed up. 

May 13 06:56:04 daos-server-0001 google_metadata_script_runner[1371]: startup-script: Created symlink from /etc/systemd/system/multi-user.target.wants/daos_agent.service to /u
May 13 06:56:04 daos-server-0001 google_metadata_script_runner[1371]: startup-script: + systemctl start daos_agent
May 13 06:56:04 daos-server-0001 google_metadata_script_runner[1371]: startup-script: + set -x
May 13 06:56:04 daos-server-0001 google_metadata_script_runner[1371]: startup-script: + echo 'BEGIN: DAOS server format'
May 13 06:56:04 daos-server-0001 google_metadata_script_runner[1371]: startup-script: BEGIN: DAOS server format
May 13 06:56:04 daos-server-0001 google_metadata_script_runner[1371]: startup-script: + [[ daos-server-0001 == \d\a\o\s\-\s\e\r\v\e\r\-\0\0\0\1 ]]
May 13 06:56:04 daos-server-0001 google_metadata_script_runner[1371]: startup-script: + dmg network scan
May 13 06:56:04 daos-server-0001 google_metadata_script_runner[1371]: startup-script: + grep --fixed-strings daos-server-0001
May 13 06:56:04 daos-server-0001 google_metadata_script_runner[1371]: startup-script: ERROR: dmg: 1 host had errors
May 13 06:56:04 daos-server-0001 google_metadata_script_runner[1371]: startup-script:   daos-server-0001 the server at daos-server-0001:10001 refused the connection
May 13 06:56:04 daos-server-0001 google_metadata_script_runner[1371]: startup-script: + echo 'All DAOS Servers started'
May 13 06:56:04 daos-server-0001 google_metadata_script_runner[1371]: startup-script: All DAOS Servers started
May 13 06:56:04 daos-server-0001 google_metadata_script_runner[1371]: startup-script: + echo 'Formatting storage on servers: daos-server-0001'
May 13 06:56:04 daos-server-0001 google_metadata_script_runner[1371]: startup-script: Formatting storage on servers: daos-server-0001
May 13 06:56:04 daos-server-0001 google_metadata_script_runner[1371]: startup-script: + dmg storage format
May 13 06:56:05 daos-server-0001 google_metadata_script_runner[1371]: startup-script: Format Summary:
May 13 06:56:05 daos-server-0001 google_metadata_script_runner[1371]: startup-script:   Hosts            SCM Devices NVMe Devices
May 13 06:56:05 daos-server-0001 google_metadata_script_runner[1371]: startup-script:   -----            ----------- ------------
May 13 06:56:05 daos-server-0001 google_metadata_script_runner[1371]: startup-script:   daos-server-0001 1           2
May 13 06:56:05 daos-server-0001 google_metadata_script_runner[1371]: startup-script: + dmg system query -v
May 13 06:56:37 daos-server-0001 google_metadata_script_runner[1371]: startup-script: Rank UUID                                 Control Address   Fault Domain      State  Reas
May 13 06:56:37 daos-server-0001 google_metadata_script_runner[1371]: startup-script: ---- ----                                 ---------------   ------------      -----  ----
May 13 06:56:37 daos-server-0001 google_metadata_script_runner[1371]: startup-script: 0    d1745607-575d-4295-b4e8-37d60e60b11e 10.128.0.54:10001 /daos-server-0001 Joined
May 13 06:56:37 daos-server-0001 google_metadata_script_runner[1371]: startup-script:
May 13 06:56:37 daos-server-0001 google_metadata_script_runner[1371]: startup-script: + echo 'Done formating DAOS server'

File: terraform/examples/daos_cluster/terraform.tfvars.perf.example (hunk 11 above)
Comment (Kmannth) [reply terraform/examples/daos_cluster/terraform.tfvars.perf.example]: I am have a patch for this.

File: terraform/examples/daos_cluster/terraform.tfvars.perf.example [hunk 19]
```diff
@@ -26,11 +26,34 @@ server_daos_scm_size       = 45
...
+#     properties = [
+#       "reclaim:disabled"
+#     ],
+#     containers = [
+#       {
+#         name = "test_container1",
+#         type = "POSIX",
+#         acls = [
+#           "A::OWNER@:rwdtTaAo",
+#           "A:G:GROUP@:rwtT"
```
Comment (Kmannth) [inline terraform/examples/daos_cluster/terraform.tfvars.perf.example]: The above 2 lines seem to permission for the Owner and Group. 

It would be much better to have the Owner and Group set in this not just their default policy.

Please make a container with both root and your user account and I think you see what I am suggesting.

File: terraform/modules/daos_server/templates/pool_cont_create.tftpl (hunk 18 above)
Comment (lsitkiew) [reply terraform/modules/daos_server/templates/pool_cont_create.tftpl]: Storage format is separated now.

**Review by markaolson — DISMISSED** [review]
After testing with and without the HPC Toolkit this PR looks good to me.

**Review by Kmannth — APPROVED** [review]
Please land this.  It was testing well for me yesterday.

**Review by markaolson — APPROVED** [review]
This PR has been tested with and without the HPC Toolkit.  It's working as intended. Excellent work @lsitkiew
